File: backend/app/models/match.py
"""Match model."""
from datetime import datetime, timezone

from ..extensions import db


class Match(db.Model):
    """Represents a single fixture between two players in a tournament."""
    __tablename__ = 'matches'

    id = db.Column(db.Integer, primary_key=True)
    tournament_id = db.Column(
        db.Integer, db.ForeignKey('tournaments.id', ondelete='CASCADE'), nullable=False
    )
    round = db.Column(db.String(50), nullable=False)  # "Round 1", "Group A", "QF", etc.
    match_order = db.Column(db.Integer, nullable=False, default=0)  # For bracket ordering

    # Players
    player1_id = db.Column(
        db.Integer, db.ForeignKey('users.id'), nullable=False
    )
    player2_id = db.Column(
        db.Integer, db.ForeignKey('users.id'), nullable=True  # Nullable for BYE
    )

    # OCR-extracted team names
    team1_name = db.Column(db.String(100), nullable=True)
    team2_name = db.Column(db.String(100), nullable=True)

    # Score
    score1 = db.Column(db.Integer, nullable=True)
    score2 = db.Column(db.Integer, nullable=True)

    # Status & verification
    status = db.Column(
        db.String(30), nullable=False, default='scheduled'
    )  # scheduled, pending_verification, verified, disputed
    reported_by = db.Column(
        db.Integer, db.ForeignKey('users.id'), nullable=True
    )
    verified_by_player1 = db.Column(db.Boolean, default=False, nullable=False)
    verified_by_player2 = db.Column(db.Boolean, default=False, nullable=False)
    admin_resolved = db.Column(db.Boolean, default=False, nullable=False)

    # Optional screenshot URL (for dispute evidence)
    screenshot_url = db.Column(db.String(500), nullable=True)

    created_at = db.Column(
        db.DateTime,
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    # Relationships
    player1 = db.relationship('User', foreign_keys=[player1_id], backref='home_matches')
    player2 = db.relationship('User', foreign_keys=[player2_id], backref='away_matches')
    reporter = db.relationship('User', foreign_keys=[reported_by])
    stats = db.relationship(
        'MatchStats', backref='match', lazy='dynamic', cascade='all, delete-orphan'
    )

    @property
    def is_bye(self):
        """Check if this match is a BYE (no opponent)."""
        return self.player2_id is None

    def to_dict(self, include_stats=False):
        """Serialize match to dictionary."""
        data = {
            'id': self.id,
            'tournament_id': self.tournament_id,
            'round': self.round,
            'match_order': self.match_order,
            'player1_id': self.player1_id,
            'player1_username': self.player1.username if self.player1 else None,
            'player2_id': self.player2_id,
            'player2_username': self.player2.username if self.player2 else None,
            'team1_name': self.team1_name,
            'team2_name': self.team2_name,
            'score1': self.score1,
            'score2': self.score2,
            'status': self.status,
            'reported_by': self.reported_by,
            'verified_by_player1': self.verified_by_player1,
            'verified_by_player2': self.verified_by_player2,
            'admin_resolved': self.admin_resolved,
            'screenshot_url': self.screenshot_url,
            'is_bye': self.is_bye,
            'created_at': self.created_at.isoformat() if self.created_at else None,
        }
        if include_stats:
            data['stats'] = [s.to_dict() for s in self.stats]
        return data

    def __repr__(self):
        return f'<Match {self.id}: {self.player1_id} vs {self.player2_id}>'

File: mobile/lib/models/match.dart
import 'user.dart';
import 'tournament.dart';

class Match {
  final int id;
  final int tournamentId;
  final String round;
  final int matchOrder;
  final int? player1Id;
  final int? player2Id;
  final String? team1Name;
  final String? team2Name;
  final int? score1;
  final int? score2;
  final String status;
  final bool verifiedByPlayer1;
  final bool verifiedByPlayer2;
  
  final User? player1;
  final User? player2;
  final Tournament? tournament;
  final List<MatchStats>? stats;

  Match({
    required this.id,
    required this.tournamentId,
    required this.round,
    required this.matchOrder,
    this.player1Id,
    this.player2Id,
    this.team1Name,
    this.team2Name,
    this.score1,
    this.score2,
    required this.status,
    required this.verifiedByPlayer1,
    required this.verifiedByPlayer2,
    this.player1,
    this.player2,
    this.tournament,
    this.stats,
  });

  factory Match.fromJson(Map<String, dynamic> json) {
    var statsList = json['stats'] as List?;
    List<MatchStats>? parsedStats;
    if (statsList != null) {
      parsedStats = statsList.map((e) => MatchStats.fromJson(e)).toList();
    }

    return Match(
      id: json['id'],
      tournamentId: json['tournament_id'],
      round: json['round'],
      matchOrder: json['match_order'],
      player1Id: json['player1_id'],
      player2Id: json['player2_id'],
      team1Name: json['team1_name'],
      team2Name: json['team2_name'],
      score1: json['score1'],
      score2: json['score2'],
      status: json['status'],
      verifiedByPlayer1: json['verified_by_player1'] ?? false,
      verifiedByPlayer2: json['verified_by_player2'] ?? false,
      player1: json['player1'] != null ? User.fromJson(json['player1']) : null,
      player2: json['player2'] != null ? User.fromJson(json['player2']) : null,
      tournament: json['tournament'] != null ? Tournament.fromJson(json['tournament']) : null,
      stats: parsedStats,
    );
  }
}

class MatchStats {
  final int id;
  final int matchId;
  final int userId;
  final String teamName;
  
  final double? possession;
  final int? shots;
  final int? shotsOnTarget;
  final int? fouls;
  final int? offsides;
  final int? cornerKicks;
  final int? freeKicks;
  final int? passes;
  final int? successfulPasses;
  final int? crosses;
  final int? interceptions;
  final int? tackles;
  final int? saves;

  MatchStats({
    required this.id,
    required this.matchId,
    required this.userId,
    required this.teamName,
    this.possession,
    this.shots,
    this.shotsOnTarget,
    this.fouls,
    this.offsides,
    this.cornerKicks,
    this.freeKicks,
    this.passes,
    this.successfulPasses,
    this.crosses,
    this.interceptions,
    this.tackles,
    this.saves,
  });

  factory MatchStats.fromJson(Map<String, dynamic> json) {
    return MatchStats(
      id: json['id'],
      matchId: json['match_id'],
      userId: json['user_id'],
      teamName: json['team_name'],
      possession: json['possession'] != null ? (json['possession'] as num).toDouble() : null,
      shots: json['shots'],
      shotsOnTarget: json['shots_on_target'],
      fouls: json['fouls'],
      offsides: json['offsides'],
      cornerKicks: json['corner_kicks'],
      freeKicks: json['free_kicks'],
      passes: json['passes'],
      successfulPasses: json['successful_passes'],
      crosses: json['crosses'],
      interceptions: json['interceptions'],
      tackles: json['tackles'],
      saves: json['saves'],
    );
  }
}
